### PR TITLE
Support multi-process shared file sets

### PR DESCRIPTION
**Pull request by nblumhardt**
Builds on https://github.com/serilog/serilog-sinks-file/pull/14 to support multi-process scenarios.

### CODE REVIEW

File: src/Serilog.Sinks.RollingFile/Sinks/RollingFile/RollingFileSink.cs [hunk 1]
```diff
@@ -54,28 +55,36 @@ public sealed class RollingFileSink : ILogEventSink, IDisposable
...
+            if (shared)
+                throw new NotSupportedException("File sharing is not supported on this platform.");
+#endif
+
             _roller = new TemplatedPathRoller(pathFormat);
             _textFormatter = textFormatter;
             _fileSizeLimitBytes = fileSizeLimitBytes;
             _retainedFileCountLimit = retainedFileCountLimit;
-            _encoding = encoding ?? new UTF8Encoding(encoderShouldEmitUTF8Identifier: false);
+            _encoding = encoding;
```
Comment (nblumhardt) [inline src/Serilog.Sinks.RollingFile/Sinks/RollingFile/RollingFileSink.cs:85]: DRY: Serilog.Sinks.File configures the default encoding, including it here means the two may drift if one is updated without updating the other.

File: src/Serilog.Sinks.RollingFile/Sinks/RollingFile/RollingFileSink.cs [hunk 2]
```diff
@@ -39,11 +39,12 @@ public sealed class RollingFileSink : ILogEventSink, IDisposable
         readonly int? _retainedFileCountLimit;
         readonly Encoding _encoding;
         readonly bool _buffered;
+        readonly bool _shared;
         readonly object _syncRoot = new object();
 
         bool _isDisposed;
         DateTime? _nextCheckpoint;
-        FileSink _currentFile;
+        ILogEventSink _currentFile;
```
Comment (nblumhardt) [inline src/Serilog.Sinks.RollingFile/Sinks/RollingFile/RollingFileSink.cs:47]: Ideally the two file sinks will expose a common interface/ABC; for now `ILogEventSink` is the best I could do here.

File: test/Serilog.Sinks.RollingFile.Tests/RollingFileLoggerConfigurationExtensionsTests.cs [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+    {
+        [Fact]
+        public void BuffferingIsNotAvailableWhenSharingEnabled()
+        {
+            Assert.Throws<ArgumentException>(() => 
+                new LoggerConfiguration()   
+                    .WriteTo.RollingFile("logs", buffered: true, shared: true));
+        }
+    }
+}
```
Comment (nblumhardt) [inline test/Serilog.Sinks.RollingFile.Tests/RollingFileLoggerConfigurationExtensionsTests.cs:16]: Tests are admittedly a bit light, it's fair to assume the `SharedFileSink` does the heavy-lifting and is tested in its own repository, but some more coverage would be nice to add here at some point too.

File: src/Serilog.Sinks.RollingFile/project.json [hunk 4]
```diff
@@ -9,16 +9,17 @@
     "iconUrl": "http://serilog.net/images/serilog-sink-nuget.png"
   },
   "dependencies": {
-    "Serilog": "2.0.0",
-    "Serilog.Sinks.File": "2.0.0"
+    "Serilog.Sinks.File": "3.0.0-dev-00735"
```
Comment (merbla) [inline src/Serilog.Sinks.RollingFile/project.json:12]: Maybe make a note to switch this back to 3.0 when this goes `dev` -> `master`

File: src/Serilog.Sinks.RollingFile/RollingFileLoggerConfigurationExtensions.cs [hunk 5]
```diff
@@ -102,11 +105,19 @@ public static class RollingFileLoggerConfigurationExtensions
...
             LoggingLevelSwitch levelSwitch = null,
-            bool buffered = false)
+            bool buffered = false,
+            bool shared = false)
         {
             if (sinkConfiguration == null) throw new ArgumentNullException(nameof(sinkConfiguration));
             if (formatter == null) throw new ArgumentNullException(nameof(formatter));
-            var sink = new RollingFileSink(pathFormat, formatter, fileSizeLimitBytes, retainedFileCountLimit, buffered: buffered);
+
+            if (shared)
```
Comment (merbla) [inline src/Serilog.Sinks.RollingFile/RollingFileLoggerConfigurationExtensions.cs]: `if(shared && buffered)` ?

Comment (nblumhardt) [inline src/Serilog.Sinks.RollingFile/RollingFileLoggerConfigurationExtensions.cs]: Oops .. copy-pasta, thanks :-)

File: src/Serilog.Sinks.RollingFile/project.json (hunk 4 above)
Comment (nblumhardt) [inline src/Serilog.Sinks.RollingFile/project.json:12]: `dotnet pack` will (should) fail on this, so if we forget it, it'll be caught by CI on the dev-master PR :+1:

**Pull request merged**